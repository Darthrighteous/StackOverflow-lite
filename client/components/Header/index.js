import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import ButtonPrimary from '../Common/ButtonPrimary';
import verifyToken from '../../utils/verifyToken';
import Toaster from '../../utils/Toaster';

/**
 * Header component
 */
class Header extends Component {
  constructor(props) {
    super(props);

    this.handleLogOut = this.handleLogOut.bind(this);
  }

  componentDidMount() {
    const { setLoggedIn } = this.props;
    if(verifyToken()) {
      setLoggedIn(true);
    }
  }

  handleLogOut() {
    const { setLoggedIn } = this.props;
    localStorage.removeItem('user');

    setLoggedIn(false);
    Toaster.success('successfully logged out', 'Goodbye');
  }

  render() {
    const { loginStatus } = this.props;
    const user = JSON.parse(localStorage.getItem('user'));
    return (
      <header>
        <div className="header-content">
          <div className="title-search-cont">
            <Link to="/">
              <div className="title">
                <i className="fab fa-stack-overflow" />
                <span>stackoverflow-lite</span>
              </div>
            </Link>
            
            <form className="search-form">
              <Link to="/search">
                <i className="fas fa-search" />
              </Link>
              <input
                id="search-bar"
                type="text"
                name="search"
                placeholder="Search all questions..."
              />
            </form>
          </div>

          <div className="user-options-cont">
            {(loginStatus) ? (
              <div className="user-options">
                <Link to={`/profile/${user.username}`}>
                  <i className="fas fa-user-circle" />
                </Link>
                <ButtonPrimary
                  label="Log Out"
                  colorTxt="#3275c3"
                  colorBtn="#ebf4ff"
                  onClick={this.handleLogOut}
                />
              </div>
              ) : (
                <div className="guest-options">
                  <ButtonPrimary
                    link="/login"
                    label="Log In"
                    colorTxt="#3275c3"
                    colorBtn="#ebf4ff"
                  />
                  <ButtonPrimary
                    link="/signup"
                    label="Sign Up"
                    colorTxt="#fff"
                    colorBtn="#3275c3"
                  />
                </div>
              )
            }
          </div>
        </div>
      </header>
    );
  }
}

Header.propTypes = {
  loginStatus: PropTypes.bool.isRequired,
  setLoggedIn: PropTypes.func.isRequired,
};

export default Header;
