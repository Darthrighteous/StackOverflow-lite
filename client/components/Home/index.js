import React from 'react';

const Home = () => (
  <div>
    <h1>StackOverflow-Lite Home</h1>
  </div>
);

export default Home;