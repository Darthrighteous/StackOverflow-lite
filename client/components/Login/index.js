import React from 'react';

const Login = () => (
  <div>
    <h1>StackOverflow-Lite Login</h1>
  </div>
);

export default Login;