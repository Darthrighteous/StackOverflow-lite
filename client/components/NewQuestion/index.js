import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { postQuestion } from '../../actions/questionActions';

/**
 * New question Component
 */
export class NewQuestion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      body: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this. handleSubmit.bind(this);
    this.handleCancel = this.handleCancel.bind(this);
  }

  handleChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const { title, body } = this.state;
    const { history, postQuestion } = this.props;

    const questionInput = { title, body };
    postQuestion(questionInput, history);
  }

  handleCancel() {
    const { history } = this.props;
    const { title, body } = this.state;

    if (title.trim() === '' && body.trim() === '') {
      return history.goBack();
    }

    this.setState({
      title: '',
      body: '',
    }); 
  }

  render() {
    const { title, body } =  this.state;
    const cancelLabel = title.trim() === '' && body.trim() === '' ? 'Back' : 'Clear';
    return (
      <div className="main-content">
        <div className="content-header">
          <h2>New Question</h2>
        </div>

        <form className="new-question-form" onSubmit={this.handleSubmit}>
          <label htmlFor="title">
            <span>Title</span>
            <input 
              id="title"
              name="title"
              placeholder="add title here"
              autoComplete="off"
              required
              value={title}
              onChange={this.handleChange}
            />
          </label>

          <label htmlFor="body">
            <span>Body</span>
            <textarea 
              id="body"
              name="body"
              placeholder="add body here"
              autoComplete="off"
              required
              value={body}
              onChange={this.handleChange}
            />
          </label>

          <div className="save-options">
            <button onClick={this.handleCancel} className="btn-cancel" type="button">{cancelLabel}</button>
            <button className="btn-post" type="submit">Post</button>
          </div>
        </form>
      </div>
    );
  }
}

NewQuestion.propTypes = {
  history: PropTypes.object.isRequired,
  postQuestion: PropTypes.func.isRequired,
};

const mapActionsToProps = {
  postQuestion
};

export default connect(null, mapActionsToProps)(NewQuestion);
