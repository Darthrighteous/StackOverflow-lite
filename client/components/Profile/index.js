import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import ContentOptions from '../Common/ContentOptions';
import ButtonNewQuestion from '../Common/ButtonNewQuestion';
import { fetchUserQuestions } from '../../actions/questionActions';
import QuestionSummary from '../Common/QuestionSummary';
import { fetchSingleUser } from '../../actions/userActions';

/**
 * Profile page component
 */
export class Profile extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sort: 'most recent',
    };

    this.setSort = this.setSort.bind(this);
  }

  componentDidMount() {
    const { fetchUserQuestions, fetchSingleUser, match } = this.props;
    fetchSingleUser(match.params.username);
    fetchUserQuestions(match.params.username);
  }
  
  setSort(sort) {
    this.setState(prevState => {
      if(prevState.sort != sort) {
        return {
          sort
        };
      }
    });
  }

  render() {
    const { sort } = this.state;
    const { userQuestions, profileOwner } = this.props;

    const signedInUser = JSON.parse(localStorage.getItem('user'));
    let isOwnProfile = false;
    if (signedInUser) {
      isOwnProfile = signedInUser.username === profileOwner.username;
    }

    const questionsLabel = isOwnProfile ? 'Your Questions' : `Questions by ${profileOwner.username} `;

    const fullName = `${profileOwner.firstname} ${profileOwner.lastname}`;
    return(
      <div className="main-content">
        {(Object.keys(profileOwner).length > 0) ? (
          <div className="profile-content">
            <div className="profile-primary">
              <ContentOptions
                title={questionsLabel}
                sort={sort}
                onSortClick={this.setSort}
              />
              {(userQuestions.length < 1) ? (
                <div className="no-resource-message">
                  <span>
                    No questions here yet
                  </span>
                </div>
              ) : (
                userQuestions.map((question) => (
                  <QuestionSummary
                    key={question.id}
                    question={question}
                    history={history}
                  />))
                )}
            </div>
            <div className="profile-side">
              <div className="user-details">
                <i className="fas fa-user-circle" />
                <div className="user-name-container">
                  <span id="fullname">{fullName}</span>
                  <span id="username">{profileOwner.username}</span>
                </div>
              </div>

              <div className="btn-new-question-cont">
                <ButtonNewQuestion />
              </div>

              <div className="user-stats">
                <div className="stats-asked">
                  <h5>Asked</h5>
                  <span id="question-count">{profileOwner.question_count}</span>
                </div>
                <div className="stats-answered">
                  <h5>Answered</h5>
                  <span id="answer-count">{profileOwner.answer_count}</span>
                </div>
                <div className="stats-comments">
                  <h5>Comments</h5>
                  <span id="comment-count">{profileOwner.comment_count}</span>
                </div>
              </div>
            </div>
          </div>
        ) : (<div />)}      
      </div>
    );
  }
}

Profile.propTypes = {
  userQuestions: PropTypes.array.isRequired,
  profileOwner: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  fetchSingleUser: PropTypes.func.isRequired,
  fetchUserQuestions: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  profileOwner: state.user,
  userQuestions: state.userQuestions,
});

const mapActionsToProps = {
  fetchSingleUser,
  fetchUserQuestions,
};

export default connect(mapStateToProps, mapActionsToProps)(Profile);
