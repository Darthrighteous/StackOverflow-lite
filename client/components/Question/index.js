import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import ButtonNewQuestion from '../Common/ButtonNewQuestion';
import Post from './Post';
import {
  fetchSingleQuestion,
  postComment,
  postAnswer,
  deletePost,
  modifyPost,
} from '../../actions/questionActions';
import getUserDetails from '../../utils/getUserDetails';
import ContentOptions from '../Common/ContentOptions';

/**
 * Question page component
 */
export class Question extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sort: 'most recent',
      isDropdownOpen: false,
      answerInput: '',
    };

    this.handleCommentPost = this.handleCommentPost.bind(this);
    this.handleAnswerInput = this.handleAnswerInput.bind(this);
    this.handleAnswerPost = this.handleAnswerPost.bind(this);
    this.handleModifyPost = this.handleModifyPost.bind(this);
    this.handleDeletePost = this.handleDeletePost.bind(this);

    this.openDropdown = this.openDropdown.bind(this);
    this.closeDropdown = this.closeDropdown.bind(this);
    this.setSort = this.setSort.bind(this);
  }

  componentDidMount() {
    const { fetchSingle, match } = this.props;
    fetchSingle(match.params.id);
  }


  setSort(sort) {
    this.setState(prevState => {
      if(prevState.sort != sort) {
        return {
          sort
        };
      }
    });
  }

  openDropdown() {
    this.setState({
      isDropdownOpen: true
    }, () => {
      document.addEventListener('click', this.closeDropdown);
    });
  }

  closeDropdown() {
    this.setState({
      isDropdownOpen: false
    }, () => {
      document.removeEventListener('click', this.closeDropdown);
    });
  }

  handleCommentPost(type, id, e) {
    e.preventDefault();
    const { postComment, history } = this.props;
    const commentInput = e.target.querySelector('input').value;
    const body = {
      body: commentInput,
    };
    postComment(type, id, body, history);
  }

  handleAnswerInput(e) {
    this.setState({
      answerInput: e.target.value,
    });
  }

  handleAnswerPost() {
    const { postAnswer, match, history } = this.props;
    const { answerInput } = this.state;
    const body = {
      body: answerInput
    };
    postAnswer(match.params.id, body, history);
  }

  handleDeletePost(resource, answerId) {
    const { deletePost, match, history } = this.props;
    const params = {
      questionId: +match.params.id,
      answerId,
    };

    deletePost(resource, params, history);
  }

  handleModifyPost(type, resourceType, answerId, editInput, undo) {
    const { modifyPost, match, history } = this.props;
    const params = {
      questionId: +match.params.id,
      answerId,
    };
    modifyPost(type, resourceType, params, history, editInput, undo);
  }
  
  render() {
    const { sort, isDropdownOpen, answerInput } = this.state;
    const { question, comments, answers, isLoggedIn } = this.props;
    const username = getUserDetails() ? getUserDetails().username : '';
    const isQuestionOwner = username === question.username ? true : false;
    return (
      <div className="main-content">

        <div className="content-header">
          <h2>{question.title}</h2>
          <ButtonNewQuestion />
        </div>

        <div className="question-answers-container">
          <div className="question-body">
            <Post
              type="questions"
              isQuestionOwner={isQuestionOwner}
              isLoggedIn={isLoggedIn}
              post={question}
              comments={comments}
              onCommentPost={this.handleCommentPost}
              onDeletePost={this.handleDeletePost}
              onModifyPost={this.handleModifyPost}
            />
          </div>

          <ContentOptions
            title="Answers"
            background
            sort={sort}
            dropdownOpenState={isDropdownOpen}
            onDropdownClick={this.openDropdown}
            onSortClick={this.setSort}
          />

          <div className="answers">
            <div className="answers-list">
              {(answers.length < 1) ? (
                <div className="no-resource-message">
                  <span>
                    No answers yet, be the first one to post an answer
                  </span>
                </div>
              ) : (
                answers.map((answer) => {
                  const comments = answer.commentlist[0].id === null
                    ? []
                    : answer.commentlist;
                  return (
                    <Post
                      key={answer.id}
                      type="answers"
                      isQuestionOwner={isQuestionOwner}
                      isLoggedIn={isLoggedIn}
                      post={answer}
                      comments={comments}
                      onCommentPost={this.handleCommentPost}
                      onDeletePost={this.handleDeletePost}
                      onModifyPost={this.handleModifyPost}
                    />);
                })
              )}
            </div>
            {(isLoggedIn) ? (
              <div className="new-answer">
                <h3>Your Answer</h3>
                <textarea className="answer-body" onChange={this.handleAnswerInput} value={answerInput} rows="10" placeholder="Type your answer here..." />
                <button type="button" onClick={this.handleAnswerPost} className="btn-post-answer">
                  Post Answer
                </button>
              </div>
            ) : (
              <div className="no-answers-message">
                <span>
                  Please&nbsp;
                  <Link to="/login">login</Link>
                  &nbsp;or&nbsp;
                  <Link to="/signup">signup</Link>
                  &nbsp;to post an answer.
                </span>
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

Question.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  question: PropTypes.object.isRequired,
  comments: PropTypes.array.isRequired,
  answers: PropTypes.array.isRequired,
  fetchSingle: PropTypes.func.isRequired,
  postComment: PropTypes.func.isRequired,
  postAnswer: PropTypes.func.isRequired,
  modifyPost: PropTypes.func.isRequired,
  deletePost: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  isLoggedIn: state.global.isLoggedIn,
  question: state.singleQuestion.question,
  comments: state.singleQuestion.comments,
  answers: state.singleQuestion.answers,
});

const mapActionsToProps = {
  fetchSingle: fetchSingleQuestion,
  postComment,
  postAnswer,
  deletePost,
  modifyPost,
};

export default connect(mapStateToProps, mapActionsToProps)(Question);