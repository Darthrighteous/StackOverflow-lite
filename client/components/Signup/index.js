import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { signUpAction } from '../../actions/authActions';

/**
 * Sign up component
 */
export class Signup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstname: '',
      lastname : '',
      email: '',
      username: '',
      password: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleSubmit(e) {
    e.preventDefault();
    const { signUpRequest, history } = this.props;
    const { firstname, lastname, email, username, password } = this.state;
    const body = {
      firstname,
      lastname,
      email,
      username,
      password,
    };
    signUpRequest(body, history);
  }

  handleChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  render() {
    const { firstname, lastname, email, username, password } = this.state;
    return (
      <div className="main-content">
        <div className="auth-form-container">	
          <div className="header" />

          <div className="body">
            <form className="auth-form" onSubmit={this.handleSubmit}>	
              <input
                required
                onChange={this.handleChange}
                type="text"
                name="firstname"
                value={firstname}
                placeholder="First Name"
              />
              <br />
              <input
                required
                onChange={this.handleChange}
                type="text"
                name="lastname"
                value={lastname}
                placeholder="Last Name"
              />
              <br />
              <input
                required
                onChange={this.handleChange}
                type="text"
                name="email"
                value={email}
                placeholder="Email"
              />
              <br />
              <input
                required
                onChange={this.handleChange}
                type="text"
                name="username"
                value={username}
                placeholder="Display Name"
              />
              <br />
              <input
                required
                onChange={this.handleChange}
                type="Password"
                name="password"
                value={password}
                placeholder="Password"
              />
              <br />
              <input
                type="submit"
                name="signUpButton"
                value="Sign Up"
              />
              <br />
              <p>
                By creating an account you agree to our&nbsp;
                <Link to="/terms-and-privacy">
                  Terms & Privacy.
                </Link>
              </p>
            </form>
          </div>

          <div className="footer">
            <p>
              Already have an account?&nbsp;
              <Link to="/login">
                Log in
              </Link>
            </p>		
          </div>
        </div>
      </div>
    );
  }
}

Signup.propTypes = {
  signUpRequest: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
};

const mapActionsToProps = {
  signUpRequest: signUpAction
};

export default connect(null, mapActionsToProps)(Signup);