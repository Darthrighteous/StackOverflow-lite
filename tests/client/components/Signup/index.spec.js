import React from 'react';
import { shallow } from 'enzyme';

import { Signup } from '../../../../client/components/Signup';

/**
 * Function to set up
 * @returns {object} - object containing wrapper
 */
function setup() {
  const props = {
    signUpRequest: jest.fn(),
    history: {
      push: jest.fn(),
    },
  };

  const signupWrapper = shallow(<Signup {...props} />);

  return {
    signupWrapper
  };
}


describe('Signup snapshot test', () => {
  const { signupWrapper } = setup();
  test('snapshot', () => {
    expect(signupWrapper).toMatchSnapshot();
  });
});

describe('event simulations', () => {
  const { signupWrapper } = setup();
  it('should handle input change', () => {
    const event = {
      target: { name: 'firstname', value: 'testname' }
    };
    signupWrapper.find('input').first().simulate('change', event);
    expect(signupWrapper.state().firstname).toEqual('testname');
  });

  it('should handle submit', () => {
    const event = {
      preventDefault: jest.fn()
    };
    signupWrapper.find('.auth-form').simulate('submit', event);
    expect(signupWrapper.state().firstname).toEqual('testname');
  });
});
