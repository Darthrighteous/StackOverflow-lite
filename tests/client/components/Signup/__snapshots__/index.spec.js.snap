// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup snapshot test snapshot 1`] = `
<div
  className="main-content"
>
  <div
    className="auth-form-container"
  >
    <div
      className="header"
    />
    <div
      className="body"
    >
      <form
        className="auth-form"
        onSubmit={[Function]}
      >
        <input
          name="firstname"
          onChange={[Function]}
          placeholder="First Name"
          required={true}
          type="text"
          value=""
        />
        <br />
        <input
          name="lastname"
          onChange={[Function]}
          placeholder="Last Name"
          required={true}
          type="text"
          value=""
        />
        <br />
        <input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          required={true}
          type="text"
          value=""
        />
        <br />
        <input
          name="username"
          onChange={[Function]}
          placeholder="Display Name"
          required={true}
          type="text"
          value=""
        />
        <br />
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="Password"
          value=""
        />
        <br />
        <input
          name="signUpButton"
          type="submit"
          value="Sign Up"
        />
        <br />
        <p>
          By creating an account you agree to our 
          <Link
            replace={false}
            to="/terms-and-privacy"
          >
            Terms & Privacy.
          </Link>
        </p>
      </form>
    </div>
    <div
      className="footer"
    >
      <p>
        Already have an account? 
        <Link
          replace={false}
          to="/login"
        >
          Log in
        </Link>
      </p>
    </div>
  </div>
</div>
`;
